File: Depth-first Search/boj_1260.py
import sys

input = sys.stdin.readline

n, m, v = map(int, input().split())

graph = [[0]*(n+1) for _ in range(n+1)]
visited = [0]*(n+1)
visited2 = [0]*(n+1)

for _ in range(m):
    a, b = map(int, input().split())
    graph[a][b] = 1
    graph[b][a] = 1

def dfs(V):
    visited[V] = 1
    print(V, end = ' ')
    for i in range(n+1):
        if graph[V][i] == 1 and visited[i] == 0:
            dfs(i)

def bfs(V):
    q = [V]
    visited2[V] = 1
    while q:
        V = q.pop(0)
        print(V, end = ' ')
        for i in range(n+1):
            if graph[V][i] == 1 and visited2[i] == 0:
                q.append(i)
                visited2[i] = 1

dfs(v)
print()
bfs(v)

File: Breadth-first Search/boj_1260.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <algorithm>
using namespace std;

void dfs(int v, int n, const vector<vector<int>>& graph) {
	stack<int> s;
	vector<int> visited(n + 1);

	s.push(v);

	while (!s.empty()) {
		v = s.top();
		s.pop();

		if (!visited[v]) {
			visited[v] = 1;
			cout << v << " ";
		}

		for (int i = graph[v].size() - 1; i >= 0; --i) {
			int nv = graph[v][i];

			if (!visited[nv]) {
				s.push(nv);
			}
		}
	}
}

void bfs(int v, int n, const vector<vector<int>>& graph) {
	queue<int> q;
	vector<int> visited(n + 1);

	q.push(v);
	visited[v] = 1;

	while (!q.empty()) {
		v = q.front();
		cout << v << " ";
		q.pop();

		for (int nv : graph[v]) {
			if (!visited[nv]) {
				q.push(nv);
				visited[nv] = 1;
			}
		}
	}
}

int main() {
	int n, m, v;

	cin >> n >> m >> v;

	vector<vector<int>> graph(n+1);
	int v1, v2;

	while (m--) {
		cin >> v1 >> v2;
		
		graph[v1].emplace_back(v2);
		graph[v2].emplace_back(v1);
	}

	for (int i = 0; i <= n; i++)
		sort(graph[i].begin(), graph[i].end());

	dfs(v, n, graph);
	cout << "\n";
	bfs(v, n, graph);

	return 0;
}
